### DISCUSSION TITLE
Static field disposed of on first lambda invocation

### DISCUSSION THREAD

**Discussion by SamuelCox** [aws/aws-lambda-dotnet#1740] [Q&A] [answered]
Some context to this before I give the example:
AWS best practice tells you to do as much as you can in the initialization code, outside of the main function handler as during the Lambda Init phase, the function gets the max amount of CPU allocated to it.

See also these docs: https://docs.aws.amazon.com/lambda/latest/operatorguide/static-initialization.html

As the docs only show this for python, this is how most people seem to do the equivalent in C#:
```
public class SecretsManagerClientProvider : ISecretsManagerClientProvider
{
    private static readonly AmazonSecretsManagerClient Client = new(new EnvironmentVariablesAWSCredentials());

    public AmazonSecretsManagerClient GetClient() => Client;
}
```

So typically from what I've seen the pattern is having a static field (often times a Dictionary for a in memory cache), or in this case it is an aws sdk client, and as we are connecting to the same account with the same credentials it shouldn't need to be instantiated per lambda execution.

Unfortunately, we found an issue where this static Client field was being disposed of, so it would only work once per cold start, and every subsequent invocation it would already have been disposed and fail.

Now, I think I know what we did wrong here but as I said the docs are only for python so I want to double check:

Even though fields can be shared across warm starts, there is no guarantee. Which I guess means we need to check if its null each time, and instantiate it again just in case? i.e
```
public class SecretsManagerClientProvider
{
    private static AmazonSecretsManagerClient Client = new(new EnvironmentVariablesAWSCredentials());

    public object GetClient()
    {
        if (Client == null) 
        {
            Client = new();
        }
        return Client;
    }
}

**Comment by normj** [comment, marked as answer]
I'm missing what is disposing the service client? There is nothing in the Lambda programming model that would trigger a dispose.

To be clear you don't have to do everything static. In .NET the Init phase includes running the constructor of the type that contains the function handler.

**Reply by SamuelCox** [reply]
Ok fair enough, our lambda is quite large and complicated, so it may be our dev team have a using statement or a dispose somewhere, ill ask them to look into it. Thanks
